File: be/services/device_service.go
package services

import (
	"errors"
	"iot/internal/model"
	"iot/internal/repository"
)

type DeviceServiceInterface interface {
	CreateDevice(name, status string) error
	GetAllDevices() ([]model.Device, error)
	GetDeviceByID(id uint) (*model.Device, error)
	GetDeviceByDeviceID(deviceID string) (*model.Device, error)
	UpdateDeviceStatus(deviceID string, status string) error
	DeleteDevice(id uint) error
}

type DeviceService struct {
	deviceRepo repository.DeviceRepositoryInterface
}

func NewDeviceService(deviceRepo repository.DeviceRepositoryInterface) DeviceServiceInterface {
	return &DeviceService{deviceRepo: deviceRepo}
}

func (s *DeviceService) CreateDevice(name, status string) error {
	if name == "" {
		return errors.New("invalid name")
	}

	if status == "" {
		return errors.New("invalid status")
	}
	return s.deviceRepo.Create(&model.Device{
		Name:   name,
		Status: status,
	})
}

func (s *DeviceService) GetAllDevices() ([]model.Device, error) {
	return s.deviceRepo.GetAll()
}

func (s *DeviceService) GetDeviceByID(id uint) (*model.Device, error) {
	return s.deviceRepo.GetByID(id)
}

func (s *DeviceService) GetDeviceByDeviceID(deviceID string) (*model.Device, error) {
	return s.deviceRepo.GetByDeviceID(deviceID)
}

func (s *DeviceService) UpdateDeviceStatus(deviceID string, status string) error {
	if status != "ON" && status != "OFF" {
		return errors.New("invalid status, must be ON or OFF")
	}
	device, err := s.deviceRepo.GetByDeviceID(deviceID)
	if err != nil {
		return errors.New("device not found")
	}

	if device == nil {
		return errors.New("device not found")
	}

	return s.deviceRepo.UpdateStatus(deviceID, status)
}

func (s *DeviceService) DeleteDevice(id uint) error {
	device, err := s.deviceRepo.GetByID(id)
	if err != nil {
		return errors.New("device not found")
	}

	if device == nil {
		return errors.New("device not found")
	}

	return s.deviceRepo.Delete(id)
}

File: fe/src/services/deviceService.ts
import { apiClient } from './apiClient';
import type {
  Device,
  CreateDeviceRequest,
  UpdateDeviceStatusRequest,
  ApiResponse,
} from '../types';

export const deviceService = {
  // Get all devices
  async getDevices(): Promise<Device[]> {
    const response = await apiClient.get<Device[]>('/devices');
    return response.data;
  },

  // Get device by ID
  async getDeviceById(id: number): Promise<Device> {
    const response = await apiClient.get<Device>(`/devices/${id}`);
    return response.data;
  },

  // Get device by device ID
  async getDeviceByDeviceId(deviceId: string): Promise<Device> {
    const response = await apiClient.get<Device>(`/devices/device/${deviceId}`);
    return response.data;
  },

  // Create new device
  async createDevice(device: CreateDeviceRequest): Promise<ApiResponse<Device>> {
    const response = await apiClient.post<ApiResponse<Device>>('/devices', device);
    return response.data;
  },

  // Update device status
  async updateDeviceStatus(deviceId: string, status: UpdateDeviceStatusRequest): Promise<ApiResponse<Device>> {
    const response = await apiClient.put<ApiResponse<Device>>(`/devices/device/${deviceId}/status`, status);
    return response.data;
  },

  // Delete device
  async deleteDevice(id: number): Promise<ApiResponse<any>> {
    const response = await apiClient.delete<ApiResponse<any>>(`/devices/${id}`);
    return response.data;
  },
};